### DISCUSSION TITLE
Recommended approach for tracking async setter state

### DISCUSSION THREAD

**Discussion by maikeriva** [pmndrs/jotai#1970] [Q&A]
Hello @dai-shi, thanks for creating Jotai! As suggested, I am opening a new issue about tracking async setter states.

I am using an async setter, and I would like to track its execution state to disable a button while executing, and an error if something gets thrown. With async atom getters I can use Suspense, but apparently not with async setters (and I'm not sure it would be a good idea even if I could)

Here's an example of what I thought would work with Suspense:
```js
const myAtomSetter = atom(null, async (get,set) => {
    await someAsyncFn()
})

const Content = () => {
    const [,setter] = useAtom(myAtomSetter)
    
    return (
        <button onClick={() => setter()}>Press me</button>
    )
}

const App = () => {
    return (
      <ErrorBoundary fallback={<p>Error</p>}>
        <Suspense fallback={<p>Loading...</p>}>
          <Content />
        </Suspense>
      </ErrorBoundary>
    )
}
```

Since it does not work, I am using the following tried and tested manual approach, which however requires writing some more code and adding some extra state:

```js
const myAtomSetter = atom(null, async (get,set) => {
    await someAsyncFn()
})

const Content = () => {
    const [isLoading, setIsLoading] = useState(false)
    const [error, setError = useState(null)
    const [,setter] = useAtom(myAtomSetter)

    const onClick = () => {
        setIsLoading(true)
        setter()
            .catch((e) => setError(e))
            .finally(() => setIsLoading(false))
    }
    
    return (
        <button disabled={isLoading} onClick={() => onClick()}>Press me</button>
    )
}

const App = () => {
    return (
        <Content />
    )
}
```

But I think there must be a more streamlined way to do it, as this is a common case. Maybe the `loadable()` utility could help? Could it be used with a setter?

Thanks a lot for your work!

_Originally posted by @maikeriva in https://github.com/pmndrs/jotai/issues/204#issuecomment-1568510128_

**Comment by dai-shi** [comment]
The recommendation is to use a loading atom
```js
const loadingAtom = atom(false)
const myAtomSetter = atom(null, async (get,set) => {
    set(loadingAtom, true)
    await someAsyncFn()
    set(loadingAtom, false)
})
```

If you want suspense, you can set a promise as the atom value.

https://github.com/pmndrs/jotai/blob/7a5ce17e38f7d5c895a1d8e9e52226881b32da67/docs/guides/async.mdx?plain=1#L79-L93
